### DISCUSSION TITLE
MYSQL Vitess Error 2027 HY000 when inserting large blob data

### DISCUSSION THREAD

**Discussion by shincurry** [planetscale/discussion#644] [Bug Reports] [answered]
Unable to run `INSERT` sql statement with large blob data (20MB+).

**The same CODE, SQL and DATA** run normally on `mysql-8.0.33`, but an error will be encountered on planetscale `mysql-8.0.34-Vitess`.

any large blob data, for example:

```typescript
// nodejs 20
// sequelize 6.37.3
// mysql2 3.10.0

const payload: any = {
    documentName: "debug:test-large-blob",
    state: (() => {
      // Create a 40MB buffer
      const bufferSize = 40 * 1024 * 1024; // 40MB in bytes
      const largeBuffer = Buffer.alloc(bufferSize);

      // Fill the buffer with some data (optional)
      largeBuffer.fill('A');

      return largeBuffer;
    })(),
  }

  await Document.upsert({
    name: payload.documentName,
    data: payload.state,
  })

```

```sql
CREATE TABLE `Document` (
  `name` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `data` longblob NOT NULL,
  `createdAt` datetime NOT NULL,
  `updatedAt` datetime NOT NULL,
  PRIMARY KEY (`name`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

```
Error: decoding parameter value failed: 6165
      at Packet.asError (/app/node_modules/mysql2/lib/packets/packet.js:728:17)
      at Execute.execute (/app/node_modules/mysql2/lib/commands/command.js:29:26)
      at Connection.handlePacket (/app/node_modules/mysql2/lib/connection.js:481:34)
      at PacketParser.onPacket (/app/node_modules/mysql2/lib/connection.js:97:12)
      at PacketParser.executeStart (/app/node_modules/mysql2/lib/packet_parser.js:75:16)
      at TLSSocket.<anonymous> (/app/node_modules/mysql2/lib/connection.js:395:25)
      at TLSSocket.emit (node:events:519:28)
      at addChunk (node:internal/streams/readable:559:12)
      at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
      at Readable.push (node:internal/streams/readable:390:5)
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:191:23)
      at TLSWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: undefined,
    errno: 2027,
    sqlState: 'HY000',
    sqlMessage: 'decoding parameter value failed: 6165',
    sql: 'INSERT INTO `Document` (`name`,`data`,`createdAt`,`updatedAt`) VALUES (?,?,?,?) ON DUPLICATE KEY UPDATE `name`=VALUES(`name`),`data`=VALUES(`data`),`updatedAt`=VALUES(`updatedAt`);',
    parameters: [
      'project:<this-is-a-project-id>',
      <Buffer ....... 23887182 more bytes>,
      '2024-07-31 06:26:39',
      '2024-07-31 06:26:39'
    ]
  }
```

**Comment by orware** [comment]
@shincurry,

Were you able to sort this issue out?

One thing I would be curious about is whether a simpler query / response works fine against the PlanetScale database as one test, to ensure your TLS configuration for the connection is working ok and that the issue isn't related to the connection itself having trouble being established.

If that is working as expected, then I would be curious if you reduce the size of your data blob to 1 MB if that would work. If it does, then I would be curious how much you can then increase it before the original error you shared appears again.

You may also reach out via https://support.planetscale.com/ to receive further assistance if needed.

**Comment by shincurry** [comment]
@orware thanks for replying.

I think there is nothing wrong with the mysql connection itself. We use planetscale database in production environment and execute a lot of various query statements. There has been no problem before until some individual case user data swelled to such a large size.

I have write a test script, and i found some new errors. https://github.com/shincurry/debug-mysql-vitess-large-blob-issue

-------------

```
SUCCESS upsert on size: 1
FAILED upsert on size: 30
SUCCESS upsert on size: 15
SUCCESS upsert on size: 23
SUCCESS upsert on size: 27
SUCCESS upsert on size: 29
```

```
FAILED upsert on size: 40
```

Try to insert a record contains 40MB blob data

```
Error: resource_exhausted: grpc: received message larger than max (83886361 vs. 67108864)
    at PromiseConnection.query (/app/.yarn/berry/cache/mysql2-npm-3.11.0-49937644f6-10c0.zip/node_modules/mysql2/promise.js:94:22)
    at upsertDocument (/app/work/debug-mysql-vitess-large-blob-issue/index.js:36:49)
    at demo (/app/work/debug-mysql-vitess-large-blob-issue/index.js:56:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'ER_UNKNOWN_ERROR',
  errno: 1105,
  sql: '\n' +
    "          INSERT INTO DebugDocument (name,data) VALUES ('test-large-blob-40',X'... 83876232 more characters,
  sqlState: 'HY000',
  sqlMessage: 'resource_exhausted: grpc: received message larger than max (83886361 vs. 67108864)'
}
```


Try to insert a record contains 30MB blob data

```
Error: internal: vtgate connection error: io: read/write on closed pipe, after 1 attempts, reqid=zZJuRwkthiZOHyOef3CsIA
    at PromiseConnection.query (/app/.yarn/berry/cache/mysql2-npm-3.11.0-49937644f6-10c0.zip/node_modules/mysql2/promise.js:94:22)
    at upsertDocument (/app/work/debug-mysql-vitess-large-blob-issue/index.js:36:49)
    at demo (/app/work/debug-mysql-vitess-large-blob-issue/index.js:56:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'ER_UNKNOWN_ERROR',
  errno: 1105,
  sql: '\n' +
    "          INSERT INTO DebugDocument (name,data) VALUES ('test-large-blob-30',X'... 62904712 more characters,
  sqlState: 'HY000',
  sqlMessage: 'internal: vtgate connection error: io: read/write on closed pipe, after 1 attempts, reqid=zZJuRwkthiZOHyOef3CsIA'
}
```

**Reply by orware** [reply]
Hello @shincurry,

I appreciate the extra tests and details you shared!

It looks like your tests confirmed an upper bound here of somewhere between 29 MB and 30 MB so you would want to ensure the values you are sending to your PlanetScale database are smaller than that if you truly need to store them within your PlanetScale database.

The main limitation you are likely running into is probably related to some of the interprocess gRPC-based communication within our Vitess layer:

```
--grpc_max_message_size int                                        Maximum allowed RPC message size. Larger messages will be rejected by gRPC with the error 'exceeding the max size'. (default 16777216)
```

One of your error messages touches on this with the `67108864` message size limit and the second one mentions a `vtgate connection error` which likely only occurs after trying one of these queries as it is knocking one of your VTGates offline temporarily and once it is available again you would begin to receive the `resource_exhausted` message as in the first error example.

Since we don't generally recommend storing large blob type values directly in the database this may be a good use case to start exploring using object storage solutions such as S3 to store this data instead, which then would allow you to store just a reference to the S3 object's location in your database vs. the entire value and you can pull your data from S3 as needed.

**Comment by shincurry** [comment]
Our server decided to use LONGBLOB storage because we saw in planetscale's documentation that it supports up to 4GB LONGBLOB columns. https://planetscale.com/learn/courses/mysql-for-developers/schema/long-strings#blob-columns . The data we store in mysql BLOB columns is not immutable files like images, audio or video, but user business data compressed into binary data (similar to JSON strings compressed into binary)

So, this is not a bug, nor is it that I am using it incorrectly, but simply that planetscale does not support BLOB column storage over 30MB? It is not difficult to migrate to services like object storage, I just hope to get a clear answer. Maybe planetscale should indicate the actual supported data size in the documentation.

Anyway, thank you very much for your reply.

**Reply by orware** [reply, marked as answer]
@shincurry,

The link you shared from our course primarily covers regular MySQL's capabilities, without our extra Vitess layer included which does add some restrictions / limitations.

While MySQL itself can support large `LONGBLOB` values, within the PlanetScale environment you will experience error messages such as the ones you encountered since the Vitess layer has a maximum message size set for the communication between MySQL + Vitess and this is also important for the overall scalability of a database in the future as it grows.

The recommendation to migrate these particular column values into object storage would be the best one I can share and with that data no longer needing to be pulled / stored directly from the PlanetScale database it will make the MySQL message sizes much smaller and avoid the errors you ran into and offload that to an object storage layer instead.

While we cover other topics in our System Limits documentation page:
https://planetscale.com/docs/reference/planetscale-system-limits#recommendations-for-handling-query-limits

It doesn't look like we cover the message size limitation situations there so that could be an improvement we can make on our end to help make things clearer for our users so thank you for the suggestion!
